### ISSUE TITLE
Opening that cuts through multiple walls issue

### ISSUE THREAD

**Issue by dvrvb** [Autodesk/revit-ifc#99]
R2019
IFC2x3CV2.0 & IFC4RV1.2

issue: opening that cuts through multiple walls isn't cut everywhere.

- probably related to commit https://github.com/Autodesk/revit-ifc/commit/958482e2651a941b5a645913dd18afa147354939
- continuation of issue started at https://sourceforge.net/p/ifcexporter/discussion/general/thread/3f41ae1fa5/

At first there seems to be a Revit 2019 issue (not the exporter).
The original model is modelled in R2019.2
When I open the model in R2019.2.1 there seems to be artefacts at level 3 and 5, which were not present in 2019.2 (as the author confirmed to me).
(the original RVT & IFC are included in the attached zip)

<img width="585" alt="image" src="https://user-images.githubusercontent.com/41319048/55573143-92c1b000-5709-11e9-80df-2e9a0e249ef2.png">

When I do an IFC-export from R2019.2.1 with the current solution (branch IFC4RV):

**IFC4RV1.2**: the cuts through the 3 walls are going well, except on level 6: there the cut is not going through the 3rd wall.
on level 3 & 5: the geometry (with the artefacts) is consistent with the geometry in R2019.2.1
<img width="385" alt="image" src="https://user-images.githubusercontent.com/41319048/55573317-f3e98380-5709-11e9-896f-710b59f3fc8e.png">

**IFC2x3CV2.0**: the cuts on level 3 & 5 don't have any artefacts (?),
on level 0 / 2 / 4 / 6 the cuts are not going through the 3rd wall.
<img width="373" alt="image" src="https://user-images.githubusercontent.com/41319048/55573394-23988b80-570a-11e9-8b86-0c100aa5f324.png">

The zip is splitted in 2 volumes (and I added 'zip' at the end, otherwise I couldn't upload it): the included IFC-models ending on IFC4 and IFC2x3 are the exports that I made with R2019.2.1 and the current solution of the IFC-exporter.

[2019-04-04 opening that cuts through multiple walls.zip.001.zip](https://github.com/Autodesk/revit-ifc/files/3044596/2019-04-04.opening.that.cuts.through.multiple.walls.zip.001.zip)

[2019-04-04 opening that cuts through multiple walls.zip.002.zip](https://github.com/Autodesk/revit-ifc/files/3044597/2019-04-04.opening.that.cuts.through.multiple.walls.zip.002.zip)

Regards,
Dirk

**Comment by dvrvb** [comment]
R2020
20190412_1200(x64) - Exporter 20.0.1.2 - Alternate UI 20.0.1.2

Just an occasional export test with the out-of-the-box IFC exporter from R2020.

IFC2x3CV2.0: same result, same issue as with R2019 
IFC4RV1.2: additional issue: the walls with the cutting issues are now on another location (runaway walls).

<img width="641" alt="image" src="https://user-images.githubusercontent.com/41319048/56784175-8c49c400-67ef-11e9-9828-b4396aa68f3e.png">

**Comment by RustyMetal** [comment]
It seems to me that the problem is not with the IFC, but in the original model. I opened 190404 Facade example.rvt and I see those triangular issues with the windows already. So what I think you have to do is add a void to those window families, one that cuts beyond both faces of the wall, so that Revit can display them properly. That should fix the IFC problem too.

**Comment by dvrvb** [comment]
eh, indeed, concerning those artefacts, I already mentioned in my message: << _on level 3 & 5: the geometry (with the artefacts) is consistent with the geometry in R2019.2.1_ >>. The main issue is about the missing cuts on level 6 for IFC4RV1.2, and on levels 0/2/4/6 for IFC2x3CV2.0.

**Comment by xpt5** [comment]
> It seems to me that the problem is not with the IFC, but in the original model. I opened 190404 Facade example.rvt and I see those triangular issues with the windows already. So what I think you have to do is add a void to those window families, one that cuts beyond both faces of the wall, so that Revit can display them properly. That should fix the IFC problem too.

Hi @RustyMetal and thank you for your answer. 
The revit family contains two nested void families in a host window family that cuts through the **necessary different walls** inside the project (as in the original post [https://sourceforge.net/p/ifcexporter/discussion/general/thread/3f41ae1fa5/](url) explained). I tried even to extend the depth of the outer void so I am sure that it cuts everything beyond the exterior wall. I find strange that the  void cutting is the issue, since it works in Revit. 
Even if it was, why is it displaying correctly in some levels (IFC tests provided by @dvrvb ) and others not when the families and hosts are the same type?

**Comment by parrela** [comment]
The difference between levels probably relates to the different wall layer heights, and where the windows and openings are placed in them. But I am not sure. I tried a number of different things, like joining the wall layers to each other, placing the windows in different layers, and making the layers the same height. What does work is placing a different window family - one of the out of the box window families cuts through all three walls like butter. So I am going to file this issue ( it is actually a duplicate of another we have already ) and see if we can figure it out.

**Comment by xpt5** [comment]
> What does work is placing a different window family - one of the out of the box window families cuts through all three walls like butter

Yes it is true but we need to have a family that cuts through different wall layers and hence nesting a double void family within the main fam. 

> So I am going to file this issue ( it is actually a duplicate of another we have already ) and see if we can figure it out.

Thank you very much for your help @parrela

**Comment by xpt5** [comment]
@parrela is there any solution estimate date? I need to decide with the project team whether we prepare a B solution if this take much more time to be solved. There are some calculations to be done based on the IFC files and our time is running out. Thank you :)

**Comment by parrela** [comment]
I can't give a time range, but I'll give a nudge to the developer. First we'll have to figure out the why, and then how to fix.

**Comment by xpt5** [comment]
After playing around with this situation I discovered that the following method makes the voids cut through all walls:

1. unjoin walls
2. rejoin them
3. slightly _overlap_ walls 

If only step 1 and 2 are done, the IFC result will be the same as in the original post. 
![Revit 2019](https://user-images.githubusercontent.com/30028308/60183381-7aec5a80-9826-11e9-8d02-83e46bc703e7.png)
![Solibri IFC 2x3 Coordination View2 0](https://user-images.githubusercontent.com/30028308/60183392-7e7fe180-9826-11e9-86d1-a3e910a5ea31.png)

**Comment by xpt5** [comment]
Update:
After some more tests I noticed that the IFC is displayed correctly when the void family is nested in the **last hosting family**. I made a simple sketch to try to explain it. 

![void cut in IFC behavior](https://user-images.githubusercontent.com/30028308/60280767-e2c4a300-9903-11e9-9b98-816b0162b71f.jpg)

I hope it helps the developer finding the cause and answer to the issue.

**Comment by dvrvb** [comment]
For R2020 with 20.1.0.1: 
- export IFC4RV1.2: no more runaway walls. Adding one cut seems to solve the issue of cutting the windows on the upper level: so for IFC4RV1.2 it seems to be fixed (previous R2020 tests were done with the out-of-the-box R2020 IFC exporter).
- export IFC3CV2.0: same results as before. Except, by adding the cut at the upper level, the window cuttings at that level are also 'fixed' (similar to the IFC4RV1.2 export).

<img width="1272" alt="image" src="https://user-images.githubusercontent.com/41319048/63649372-fa21e000-c73c-11e9-9250-f352fc4901d9.png">

**Comment by xpt5** [comment]
Hi @dvrvb, thanks for the heads up!
What do you mean by "adding one cut"? An extra cutting family needs to be added so it works correctly?

P.S - sorry the delay, I have been pretty busy for the last weeks,

**Comment by dvrvb** [comment]
Hi João,
No extra cutting family. Like a butcher I just made a simple cut with the knife tool.
But I should expressed myself better, since the tool is called 'split element'.
<img width="181" alt="image" src="https://user-images.githubusercontent.com/41319048/64266308-ae3d1b00-cf34-11e9-9944-3510d0ebc28c.png">
Regards,
Dirk

**Comment by xpt5** [comment]
Ok now I understand :)
I will try to get rvt2020 myself so I can test it. We just had to rebuild the whole facade system so we could see those openings correctly in IFC. 

Regards,
João

**Comment by dvrvb** [comment]
But maybe that cut should also work in R2019: I didn't tested it in R2019.
Did you get rid of those triangular artefacts on the 3rd and the 5th level in the rvt model?

**Comment by xpt5** [comment]
Those artifacts are pretty random, they just pop up even if the walls are aligned/joined. I don't know why that was happening. I will check the old file.

Anyways, the same facade with a different window family + external double void made the trick (see drawing posted on Jun 27). Everything behaved as expected in IFC.
It is a shame though, that we had to rebuild our window family system. It was developed so we could use it together with a Dynamo script for facade optimization.

**Comment by dvrvb** [comment]
It's fair to close this issue: seems to be fixed in R2023.1 (exp 23.2.3.0) for IFC4RV.
Even better, the exporter fixed the artefacts (which are still present in the RVT model).
<img width="926" alt="2023-01-12" src="https://user-images.githubusercontent.com/41319048/212090626-747c390f-9b09-47e3-8a74-14271ca9e971.png">
